### PR TITLE
Min-SNR gamma support for Dreambooth training

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

Implements min-SNR gamma training along with the v_prediction fix, alongside prior preservation loss.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/dreambooth/train_dreambooth.py [hunk 1]
```diff
@@ -224,6 +224,30 @@ def import_model_class_from_model_name_or_path(pretrained_model_name_or_path: st
         raise ValueError(f"{model_class} is not supported.")
 
 
+def compute_snr(timesteps, noise_scheduler):
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth.py:227]: Maybe we should graduate this function to `training_utils.py` at this point as Min-SNR training is quickly becoming a gold standard. @patrickvonplaten, thoughts?

File: examples/dreambooth/train_dreambooth.py [hunk 2]
```diff
@@ -1261,17 +1292,36 @@ def compute_text_embeddings(prompt):
...
+                    snr = compute_snr(timesteps, noise_scheduler)
+                    mse_loss_weights = (
+                        torch.stack(
+                            [snr, args.snr_gamma * torch.ones_like(timesteps)],
+                            dim=1,
+                        ).min(
+                            dim=1
+                        )[0]
+                        / snr
+                    )
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth.py:1314]: Break it down like you did in the other PRs, please.

**Review by sayakpaul — APPROVED** [review]
LGTM given the feedback has been addressed :-)

**Comment by bghira** [comment]
@sayakpaul matches the SDXL trainer now.

File: examples/dreambooth/train_dreambooth.py (hunk 1 above)
Comment (bghira) [reply examples/dreambooth/train_dreambooth.py:227]: totally. especially because it's unmodified reference code

Comment (patrickvonplaten) [reply examples/dreambooth/train_dreambooth.py:227]: Yeah ok for me!

**Review by patrickvonplaten — APPROVED** [review]

**Comment by sayakpaul** [comment]
Failing test is flaky.
